### PR TITLE
feat(cb2-7842): update logic and add tests

### PR DESCRIPTION
**Pull request by naathanbrown**
## Ticket title

I broke it and it regressed. So I did some tests and fixed it.

## Checklist

- [ ] Branch is rebased against the latest develop/common
- [ ] Necessary `id` required prepended with `"test-"` have been checked with automation testers and added
- [ ] Code and UI has been tested manually after the additional changes
- [ ] PR title includes the JIRA ticket number
- [ ] Squashed commits contain the JIRA ticket number
- [ ] Link to the PR added to the repo
- [ ] Delete branch after merge

### CODE REVIEW

File: src/app/services/technical-record/technical-record.service.spec.ts [hunk 1]
```diff
@@ -378,6 +380,271 @@ describe('TechnicalRecordService', () => {
...
+          }
+        ];
+
+        vehicle.techRecord = techRec;
+
+        const res = service.calculateTechRecordToUpdate(vehicle);
+
+        expect(res).toBe(techRec[0]);
+      });
+      it('should error when there is only a archived', () => {
```
Comment (leifkemp-bjss) [inline src/app/services/technical-record/technical-record.service.spec.ts:533]:
```suggestion
      it('should error when there is only an archived', () => {
```

File: src/app/services/technical-record/technical-record.service.spec.ts [hunk 2]
```diff
@@ -378,6 +380,271 @@ describe('TechnicalRecordService', () => {
...
+
+        expect.hasAssertions();
+
+        try {
+          service.calculateTechRecordToUpdate(vehicle, StatusCodes.PROVISIONAL);
+        } catch (err: any) {
+          expect(err.message).toContain('Cannot find a provisional or current to update');
+        }
+      });
+      it('should error when there is only a archived', () => {
```
Comment (leifkemp-bjss) [inline src/app/services/technical-record/technical-record.service.spec.ts:461]:
```suggestion
      it('should error when there is only an archived', () => {
```

**Review by leifkemp-bjss — DISMISSED** [review]

File: src/app/services/technical-record/technical-record.service.ts [hunk 3]
```diff
@@ -172,6 +170,23 @@ export class TechnicalRecordService {
...
+      if (!newTechRecord) {
+        newTechRecord = newVehicleTechRecord.techRecord.find(techRecord => techRecord.statusCode === StatusCodes.PROVISIONAL);
+      }
+    }
+
+    if (!newTechRecord) throw new Error(`Cannot find a provisional or current to update`);
+
+    return newTechRecord;
+  }
+
```
Comment (gjulien-bjss) [inline src/app/services/technical-record/technical-record.service.ts:189]:
```suggestion
  calculateTechRecordToUpdate(newVehicleTechRecord: VehicleTechRecordModel, recordToArchiveStatus?: StatusCodes): TechRecordModel {
    const filteredRecords = newVehicleTechRecord.techRecord.filter(techRecord => techRecord.statusCode !== StatusCodes.ARCHIVED);
    const newTechRecord =
      filteredRecords.find(techRecord => techRecord.statusCode === (recordToArchiveStatus ?? StatusCodes.CURRENT)) ??
      filteredRecords.find(techRecord => techRecord.statusCode === StatusCodes.PROVISIONAL);

    if (!newTechRecord) throw new Error(`Cannot find a provisional or current to update`);

    return newTechRecord;
  }
  
```
Don't really know if it's nicer, but passes all the tests

File: src/app/services/technical-record/technical-record.service.ts [hunk 4]
```diff
@@ -172,6 +170,23 @@ export class TechnicalRecordService {
...
+      newTechRecord = newVehicleTechRecord.techRecord.find(techRecord => techRecord.statusCode === StatusCodes.CURRENT);
+      if (!newTechRecord) {
+        newTechRecord = newVehicleTechRecord.techRecord.find(techRecord => techRecord.statusCode === StatusCodes.PROVISIONAL);
+      }
+    }
+
+    if (!newTechRecord) throw new Error(`Cannot find a provisional or current to update`);
+
+    return newTechRecord;
+  }
```
Comment (tahayazici) [inline src/app/services/technical-record/technical-record.service.ts:188]: 1. To not hijack the order of methods in this service, can you please move this helper method under `filterTechRecordByStatusCode()`.
2. You can simplify the logic as such:
```suggestion
  calculateTechRecordToUpdate(vehicle: VehicleTechRecordModel, status?: StatusCodes): TechRecordModel {
    let techRecord = vehicle.techRecord.find(techRecord => techRecord.statusCode === StatusCodes.CURRENT);

    if (!techRecord || status === StatusCodes.PROVISIONAL) {
      techRecord = vehicle.techRecord.find(techRecord => techRecord.statusCode === StatusCodes.PROVISIONAL);
    }

    if (!techRecord) throw new Error(`Cannot find a provisional or current to update`);

    return techRecord;
  }
```

Comment (naathanbrown) [reply src/app/services/technical-record/technical-record.service.ts:188]: This is nice! Thanks man

File: src/app/services/technical-record/technical-record.service.ts (hunk 3 above)
Comment (naathanbrown) [reply src/app/services/technical-record/technical-record.service.ts:189]: Like this, but going to go with Taha's suggestion below instead, thank you!

**Review by tahayazici — APPROVED** [review]

**Review by leifkemp-bjss — APPROVED** [review]

**Review by gjulien-bjss — APPROVED** [review]

**Pull request merged**